core/src/main/java/org/radargun/stages/mapreducestream/MapReduceStreamStage.java

```diff
@@ -0,0 +1,213 @@
+      Map<Integer, Report.SlaveResult> durationsResult = new HashMap<>();
+
+      for (MapReduceStreamAck ack : Projections.instancesOf(acks, MapReduceStreamAck.class)) {
+         if (ack.stats != null) {
+            DataOperationStats opStats = (DataOperationStats) ack.stats.getOperationsStats().get(Streamable.STREAMABLE.name);
+
+            opStats.setTotalBytes((Long) masterState.get(totalBytesKey));
+            test.addStatistics(testIteration, ack.getSlaveIndex(), Collections.singletonList(ack.stats));
+            durationsResult.put(ack.getSlaveIndex(), new Report.SlaveResult(opStats.getResponseTimes(), false));
+            test.addResult(testIteration, new Report.TestResult("Map/Reduce durations", durationsResult, "", false));
+         }
+         if (ack.numberOfResultKeys != null) {
+            numberOfResultKeysResult.put(ack.getSlaveIndex(), new Report.SlaveResult(ack.numberOfResultKeys, false));
+            test.addResult(testIteration, new Report.TestResult("Key count in Map/Reduce result map",
+                    numberOfResultKeysResult, "", false));
+         }
+      }
+
+      return StageResult.SUCCESS;
+   }
+
+   @Override
+   public DistStageAck executeOnSlave() {
+      if (!isServiceRunning()) {
+         return errorResponse("Service is not runnning", null);
+      }
+
+      if (slaveState.getSlaveIndex() == 0) {
+         DistStageAck result = null;
+         Map<Object, Object> prevPayloadMap = null;
+
+         Statistics stats = new DefaultStatistics(new DataOperationStats());
+         stats.begin();
+
+         for (int i = 0; i < numExecutions; i++) {
+            prevPayloadMap = payloadMap;
+
+            result = executeMapReduceStreamTask(stats);
+
+            if (prevPayloadMap != null) {
+               if (prevPayloadMap.keySet().equals(payloadMap.keySet())
+                       && prevPayloadMap.entrySet().equals(payloadMap.entrySet())) {
+                  log.info(i + ": Got the same results for two Map/Reduce runs");
+               } else {
+                  log.error(i + ": Did not get the same results for two Map/Reduce runs");
+                  break;
+               }
+            }
+
+            if (result.isError() && exitOnFailure) {
+               break;
+            }
+         }
+         stats.end();
+
+         log.info("Executed Map reduce stream stage on slave");
+         return result;
+
+      } else {
+         return new MapReduceStreamAck(slaveState);
+      }
+   }
+
+   private DistStageAck executeMapReduceStreamTask(Statistics stats) {
+      long durationNanos;
+      long start;
+
+      configureMapReduceTask();
+
+      MapReduceStreamAck ack = new MapReduceStreamAck(slaveState);
+      try {
+         start = TimeService.nanoTime();
+
+         consumer.accept(streamable.stream(cacheName));
+
+         Object result = consumer.getResult();
+         if (result instanceof Map) {
```

Comment:
Not really elegant. If you want to result be just reported, it should be the consumer (or rather the Function) that processes it to something presentable.